### PR TITLE
Imps regain horns after some time

### PR DESCRIPTION
**Pull request by KnowOne134**
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

## What does this pull request do?

retail testing shows that Imps regain their horns at varying times of 30, 35, 40, 45, and 60 seconds which is very rare to see
uses the same random that causes the horn to break to apply the time that it will reapply it  (1,19)

## Steps to test these changes

fight an imp, with godmode to hit the crits
once horn breaks turn back to it and withon 60s it will get a new one

### CODE REVIEW

**Review by Xaver-DaRed — APPROVED** [review]

File: scripts/mixins/families/imp.lua [hunk 1]
```diff
@@ -6,9 +6,24 @@ g_mixins.families = g_mixins.families or {}
 g_mixins.families.imp = function(impMob)
 
     -- 20% chance to break horn on critical hit
+    -- Reacquires horn in stages of 30, 35, 40, 45, and a very rare 60 seconds
     impMob:addListener("CRITICAL_TAKE", "IMP_CRITICAL_TAKE", function(mob)
-        if math.random(100) < 20 and mob:getAnimationSub() == 0 then
+        local random = math.random(100)
+        if random < 20 and mob:getAnimationSub() == 0 then
             mob:setAnimationSub(1)
+            if not mob:getLocalVar("hornDisabled") == 1 then
```
Comment (Xaver-DaRed) [inline scripts/mixins/families/imp.lua]: `not` only works with true/false comparations.

I think `if mob:getLocalVar("hornDisabled") ~= 1` would work.

File: scripts/mixins/families/imp.lua [hunk 2]
```diff
@@ -6,9 +6,24 @@ g_mixins.families = g_mixins.families or {}
...
+                local time = 0
+                if random <= 2 then
+                    time = 60
+                else
+                    time = 25 + (math.ceil(random / 5) * 5)
+                end
+                mob:timer(time * 1000, function(mob)
```
Comment (Xaver-DaRed) [inline scripts/mixins/families/imp.lua]: `function(mob)` should be changed to `function(something_else_not_mob)`

File: scripts/mixins/families/imp.lua [hunk 3]
```diff
@@ -6,9 +6,24 @@ g_mixins.families = g_mixins.families or {}
...
             mob:setAnimationSub(1)
+            if mob:getLocalVar("hornDisabled") ~= 1 then
+                local time = 0
+                if random <= 2 then
+                    time = 60
+                else
+                    time = 25 + (math.ceil(random / 5) * 5)
+                end
+                mob:timer(time * 1000, function(mobArg)
+                    if mob:isAlive() then
```
Comment (jmcmorris) [inline scripts/mixins/families/imp.lua:22]:
```suggestion
                    if mobArg:isAlive() then
```

File: scripts/mixins/families/imp.lua [hunk 4]
```diff
@@ -6,9 +6,24 @@ g_mixins.families = g_mixins.families or {}
...
+                        mob:setAnimationSub(0)
```
Comment (jmcmorris) [inline scripts/mixins/families/imp.lua:23]:
```suggestion
                        mobArg:setAnimationSub(0)
```

**Review by jmcmorris — COMMENTED** [review]
I'm not 100% sure this is necessary but you don't need to use the upvalue `mob` here since you have `mobArg`.

File: scripts/mixins/families/imp.lua (hunk 3 above)
Comment (cocosolos) [reply scripts/mixins/families/imp.lua:22]: Is `mobArg` even valid? It's my understanding that `mob` is already available to the inner function without passing anything as an arg, and nothing is actually passed as an arg anyway so `mobArg` is just nil.

Comment (jmcmorris) [reply scripts/mixins/families/imp.lua:22]: Yes, mobArg is valid. mob:timer() queues an action and here's the code that handles that action. https://github.com/LandSandBoat/server/blob/base/src/map/ai/helpers/action_queue.cpp#L77

It is true that mob is available inside the anonymous function as an upvalue. My comment was mostly coming from my own experience of avoiding using pointers as upvalues since it could potentially be deleted between creation and invocation, but the reality is that mobs (and most things) are created in memory and stay there so it isn't a real problem here.
TLDR: it kind of doesn't matter :)

Comment (cocosolos) [reply scripts/mixins/families/imp.lua:22]: Ah I was looking at the struct and somehow missed the handle function, thanks for pointing that out!

Comment (WinterSolstice8) [reply scripts/mixins/families/imp.lua:22]: essentially, when timer is invoked it calls the function you made directly

```
function(mobArg)
    if mob:isAlive() then
        ...
        end
        ...
end
```

Note that `mob` isn't in scope here, this function _technically_ shouldn't know about `mob`, but it does know about `mobArg`, which will be equivalent to using `mob`  (as that's what's passed in to this function when the timer is invoked)

**Review by claywar — APPROVED** [review]
